pkg/parse/sync_status_cache.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parse
+
+import (
+	"sync"
+
+	"kpt.dev/configsync/pkg/remediator/conflict"
+	"kpt.dev/configsync/pkg/status"
+	"kpt.dev/configsync/pkg/syncer/reconcile/fight"
+)
+
+// SyncErrorCache is a collection of sync errors, locked for thread-safe use.
+type SyncErrorCache struct {
+	statusMux sync.RWMutex
+	// Errors from the Remediator & Updater
+	conflictHandler conflict.Handler
+	// Errors from the Remediator
+	fightHandler fight.Handler
+	// Errors from the Updater
+	validationErrs status.MultiError
+	applyErrs      status.MultiError
+	watchErrs      status.MultiError
+}
+
+// NewSyncErrorCache constructs a new SyncErrorCache with shared handlers
+func NewSyncErrorCache(conflictHandler conflict.Handler, fightHandler fight.Handler) *SyncErrorCache {
+	return &SyncErrorCache{
+		conflictHandler: conflictHandler,
+		fightHandler:    fightHandler,
+	}
+}
+
+// Errors returns the latest known set of errors from the updater and remediator.
+func (s *SyncErrorCache) Errors() status.MultiError {
+	s.statusMux.RLock()
+	defer s.statusMux.RUnlock()
+
+	// Ordering here is important. It needs to be the same as Updater.Update.
+	var errs status.MultiError
+	errs = status.Append(errs, s.remediatorErrors())
+	errs = status.Append(errs, s.updaterErrors())
+	return errs
+}
+
+// UpdaterErrors combines validation, apply, and watch errors.
+func (s *SyncErrorCache) UpdaterErrors() status.MultiError {
+	s.statusMux.RLock()
+	defer s.statusMux.RUnlock()
+	return s.remediatorErrors()
+}
+
+// updaterErrors combines validation, apply, and watch errors.
+func (s *SyncErrorCache) updaterErrors() status.MultiError {
```

Comment:
Does the read here need to be locked?

Reply:
Nah. This is the private method for internal use. It's only called by other public methods that handle the locking: UpdaterErrors & Errors. Errors can't call UpdaterErrors because the mutex isn't re-entrant, so it calls updaterErrors instead.

The non-locking internal method is just a code re-use strategy to avoid having the ordering of error types hard-coded in two different places.

Reply:
removed.